### PR TITLE
core(uses-rel-preconnect): add FCP and LCP savings

### PR DESCRIPTION
**Pull request by adrianaixba**

### CODE REVIEW

**Comment by adrianaixba** [comment]
might need a double check to make sure there's good coverage on the tests

File: core/audits/uses-rel-preconnect.js [hunk 1]
```diff
@@ -126,6 +127,7 @@ class UsesRelPreconnectAudit extends Audit {
     const settings = context.settings;
 
     let maxWasted = 0;
+    let maxWastedFCP = 0;
```
Comment (adamraine) [inline core/audits/uses-rel-preconnect.js:130]: nit: can we make these `maxWatedLcp` and `maxWastedFcp`

File: core/test/audits/uses-rel-preconnect-test.js [hunk 2]
```diff
@@ -14,12 +14,14 @@ const mainResource = {
...
   timing: {receiveHeadersEnd: 0.5},
   networkEndTime: 1000,
+  priority: 'High',
 };
 
 function buildArtifacts(networkRecords) {
   const trace = createTestTrace({
     timeOrigin: 0,
     largestContentfulPaint: 5000,
+    firstContentfulPaint: 5000,
```
Comment (adamraine) [inline core/test/audits/uses-rel-preconnect-test.js:24]: Can you add a test case that has LCP savings but not FCP savings? You should be able to achieve this by lowering this value to 4000 or something and then using a request that starts after FCP and ends before LCP in the test case.

File: core/test/audits/uses-rel-preconnect-test.js [hunk 3]
```diff
@@ -353,4 +354,44 @@ describe('Performance: uses-rel-preconnect audit', () => {
     assert.equal(result.score, 1);
     assert.equal(result.warnings.length, 1);
   });
+
+  it(`should have LCP savings and not FCP savings `, async () => {
```
Comment (adamraine) [inline core/test/audits/uses-rel-preconnect-test.js:358]:
```suggestion
  it('should have LCP savings and not FCP savings', async () => {
```

File: core/test/audits/uses-rel-preconnect-test.js [hunk 4]
```diff
@@ -343,4 +354,44 @@ describe('Performance: uses-rel-preconnect audit', () => {
...
+          dnsStart: 300,
+          connectStart: 350,
+          connectEnd: 400,
+        },
+        priority: 'High',
+      },
+      {
+        url: 'https://cdn.example.com/second',
+        initiator: {},
+        networkRequestTime: 4000,
```
Comment (adamraine) [inline core/test/audits/uses-rel-preconnect-test.js:375]:
```suggestion
        networkRequestTime: 4000, // starts *after* FCP
```

File: core/audits/uses-rel-preconnect.js [hunk 5]
```diff
@@ -147,6 +149,13 @@ class UsesRelPreconnectAudit extends Audit {
       if (node.type === 'network' ) lcpGraphURLs.add(node.record.url);
     });
 
+    const fcpGraph =
+      await LanternFirstContentfulPaint.getPessimisticGraph(pageGraph, processedNavigation);
+    const fcpGraphURLs = new Set();
+    fcpGraph.traverse(node => {
+      if (node.type === 'network' ) fcpGraphURLs.add(node.record.url);
```
Comment (adamraine) [inline core/audits/uses-rel-preconnect.js:156]: Nit: I realized this is copied but can you fix in both places?
```suggestion
      if (node.type === 'network') fcpGraphURLs.add(node.record.url);
```

**Review by adamraine — APPROVED** [review]

File: core/audits/uses-rel-preconnect.js (hunk 5 above)
Comment (adamraine) [reply core/audits/uses-rel-preconnect.js]: https://eslint.org/docs/latest/rules/no-multi-spaces

Looks like the lint rule only flags if there are *multiple* spaces.

**Pull request merged**